keps/sig-cluster-lifecycle/kubeadm/20190925-component-configs.md

```diff
@@ -0,0 +1,223 @@
+
+##### User supplied Kubernetes Core Component Config management
+
+As user changes need to be maintained, user supplied configs are stored config maps.  
+However, in those cases kubeadm is not responsible for config migration upon upgrade. If the new component version is unable to consume the user supplied config, kubeadm is going to bail out with an error message, prompting the user to perform a manual config migration beforehand. This is a behavior similar to that for addon configs.
+
+##### Allow Kustomize use over core Kubernetes component configs
+
+Kubeadm shall extend its current Kustomize support to include patching of core Kubernetes component configs. Currently, this will include only kubelet’s component config. However, in the future this might extend to the API Server, Scheduler and Controller Manager as well.  
+Most notably, this feature is not going to be applied to kube-proxy’s config by kubeadm as this is an addon component config and the responsibility for this is going to be with the corresponding addon operator.
+
+Kustomize patches will be applied on both user supplied and kubeadm generated core component configs, just before they are stored in the local files from which the config will be consumed by the components. In the case of user supplied core component configs, neither the Kustomize patches, nor the result after applying them is going to be stored on the cluster. The config maps are going to contain only the vanilla user supplied component configs.
+
+If a user has provided Kustomize patches for a config version, that is different from the current one, kubeadm is going to bail out with an error, prompting the user to either delete the old patches, or to upgrade them to the new version. This is best implemented using a kubeadm pre-flight check.
+
+UX wise, this feature is going to be reusing the existing kubeadm options that were introduced with the implementation of the kubeadm advanced configurations KEP. This allows for kustomizations to be applied to component configs on a per-node basis during init, join and upgrade just before the config is supposed to be stored on the local node file.
+
+### Stop using component config internal types
+
+As kubeadm will no longer default, validate or migrate component configs inside its code base, the need to use component config internal types is no longer founded. In that case the usage of internal types can be removed. This would also allow us to reduce the number of imported packages and to drop one of the last dependencies on the kubernetes/kubernetes code base.  
+The switch is going to be performed by replacing each internal type with a corresponding versioned type. This, by itself, is not going to have any impact on users.
+
+### “kubeadm upgrade plan” changes
```

Comment:
I'm not sure this is enough.
both init and join can be affected by config validation errors, and upgrade apply/upgrade node preflight as well

Reply:
That's why I initially marked this part of the KEP as optional, as I am not sure how it will turn out. I'll relabel it as "optional" again.